### PR TITLE
feat(surveys): Even distribution of Survey response collection over sampling interval

### LINKED ISSUE #25722: Adaptive response sampling for surveys
### Feature request

We currently allow %-based rollouts for surveys, as well as the option to set a response limit.

For some users with high volumes, that's not enough. Here is what we've heard:
- Being able to specify a response limit per time period, e.g. "250 responses per month", or "7 responses per day"
- Ideally, PostHog would _adapt_ the rollout of surveys across the time period specified by the user, e.g. when a user wants 250 responses over a one-month period, we make sure they receive approximately 8 responses per day

One proposal how this could be achieved from @neilkakkar: "I think the naive targeting all users and adjusting the rollout % based on how many responses have been received so far would work pretty well - and better than trying to extra optimise it based on several other factors / guestimating cohort sizes who are active.. This is also reasonably straightforward to implement - we have all the pieces in place already."

More details in this Slack thread here: https://posthog.slack.com/archives/C07QD3LT8U9/p1728406374642269

### Debug info

_No response_

### PR DESCRIPTION
**Pull request by Phanatic**
## Problem

Resolves : https://github.com/PostHog/posthog/issues/25722

We currently allow %-based rollouts for surveys, as well as the option to set a response limit.

For some users with high volumes, that's not enough. Here is what we've heard:

Being able to specify a response limit per time period, e.g. "250 responses per month", or "7 responses per day"
Ideally, PostHog would adapt the rollout of surveys across the time period specified by the user, e.g. when a user wants 250 responses over a one-month period, we make sure they receive approximately 8 responses per day

### Changes to the `Survey` Model

This PR introduces the following changes to the `Survey` model to support even distribution of feature flag rollout percentages and response sampling limits:

---

#### **New Fields**

1. **`response_sampling_daily_limits`**
   - **Type**: `JSONField`
   - **Description**: Stores the daily response limits and rollout percentages for the `response_sampling_interval`. This is a precomputed schedule, calculated before saving the survey, to guide response collection and feature flag rollout updates.
   - **Structure**:
     ```json
     [
         {"date": "2024-12-12T00:00:00+00:00", "daily_response_limit": 50, "rollout_percentage": 10.0},
         {"date": "2024-12-13T00:00:00+00:00", "daily_response_limit": 100, "rollout_percentage": 20.0},
     ]
     ```
   - **Purpose**:
     - Tracks cumulative response limits (`daily_response_limit`) for each day in the sampling interval.
     - Ensures even distribution of the `rollout_percentage` over the interval.

2. **`response_sampling_start_date`**
   - **Type**: `DateTimeField`
   - **Description**: Indicates the UTC start date for the response sampling period. This is the starting point for calculating the `response_sampling_daily_limits`.

3. **`response_sampling_limit`**
   - **Type**: `IntegerField`
   - **Description**: The total number of responses to collect over the `response_sampling_interval`. This is evenly distributed across the interval.

4. **`response_sampling_interval`**
   - **Type**: `IntegerField`
   - **Description**: Represents the number of intervals (e.g., days, weeks, months) over which responses are distributed.

5. **`response_sampling_interval_type`**
   - **Type**: `CharField`
   - **Choices**: `("day", "Day"), ("week", "Week"), ("month", "Month")`
   - **Description**: Specifies the type of interval for response sampling (e.g., daily, weekly, or monthly).

### Distributing survey response limits over a time period
``` bash
S = <total number of responses accepted> 
D = <duration>
n = <number of days in duration>
u = <per day response count>
```

If a customer wants to collect 500 responses over 30 days : 
``` 
S = 500
n = 30 days
u = Math.ceil(500/30)  = 17
```

We would evenly distribute the number of responses we collect over the time period, giving us 17 responses per day.
We would then turn this into an arithmetic progression where element in the array is the total number of responses we hope to get by that day. 

1. **Response Sampling Limits Update**
   - Introduced the `update_response_sampling_limits` function to:
     - Generate `response_sampling_daily_limits` with even distribution of rollout percentages over the `response_sampling_interval`.
     - Track cumulative `daily_response_limit` to ensure accurate total responses collected up to each day.
     - Ensure all dates are stored in UTC for consistency.
   - The schedule ensures equal distribution of rollout percentages across the interval, with any remainder evenly distributed across the first few days.

   Example `response_sampling_daily_limits`:
```json
   [
       {"date": "2024-12-12T00:00:00+00:00", "daily_response_limit": 50, "rollout_percentage": 10.0},
       {"date": "2024-12-13T00:00:00+00:00", "daily_response_limit": 100, "rollout_percentage": 20.0},
       {"date": "2024-12-14T00:00:00+00:00", "daily_response_limit": 150, "rollout_percentage": 30.0},
   ]
```

**2. Celery task for daily rollout updates**

Created a Celery task to update the rollout percentage of a linked targeting_feature_flag based on the calculated response_sampling_daily_limits.
 1. Ensures the feature flag's rollout_percentage aligns with the response_sampling_daily_limits for the current UTC date.
 2. Updates the feature flag daily to maintain an even distribution of response collection.

## Does this work well for both Cloud and self-hosted?

Yes

## How did you test this code?

**Response sampling daily limits:**
1. Verifies that response_sampling_daily_limits correctly distributes rollout percentages and cumulative response limits across the response_sampling_interval.
2. Handles intervals for days, weeks, and months.

**Celery task:**
1. Confirms that the feature flag's rollout_percentage is updated correctly based on the response_sampling_daily_limits for today in UTC.
2. Ensures smooth operation for various sampling intervals and edge cases.

### CODE REVIEW

**Review by danielbachhuber — COMMENTED** [review]
I did a read through and nothing obviously incorrect stands out to me.

How does the customer interact with the setting?

**Comment by Phanatic** [comment]
@danielbachhuber thanks! 
I'm building the UI now, this is what it would look like. Still needs some refinement as the UI elements are fighting for space. 

image

File: posthog/tasks/tasks.py [hunk 1]
```diff
@@ -799,6 +799,13 @@ def update_survey_iteration() -> None:
     update_survey_iteration()
 
 
+@shared_task(ignrore_result=True)
```
Comment (dmarticus) [inline posthog/tasks/tasks.py:802]:
```suggestion
@shared_task(ignore_result=True)
```

not sure what this decorator does, but there's a typo in `ignore_result`, so my guess is that it won't do what you want ;)

**Review by dmarticus — APPROVED** [review]
one comment, looks sensible otherwise.  cool feature!

**Pull request closed without merging**